File: bogo2bogo/TextFinder/Textfinder.cpp
#include "Textfinder.h"
#include "ui_Textfinder.h"
#include <QFile>
#include <QFileDialog>
#include <QMessageBox>
#include <QStandardPaths>
#include <QTextStream>

QFont monoFont = QFont("Consolas", 10);

TextFinder::TextFinder(QWidget *parent) : QWidget(parent), ui(new Ui::TextFinder)
{
   ui->setupUi(this);
   setWindowIcon(QIcon(":/eye.png")); // load an image
   loadTextFile();

   // setup signals & slots
   ui->textEdit->setReadOnly(true);
   ui->textEdit->setEnabled(false);
   ui->textEdit->setFont(monoFont);
   QObject::connect(ui->findButton, SIGNAL(clicked()), this, SLOT(onFindButtonClicked()));
   QObject::connect(ui->openButton, SIGNAL(clicked()), this, SLOT(onOpenButtonClicked()));
}

TextFinder::~TextFinder() { delete ui; }

void TextFinder::loadTextFile(const QString &filePath /*=""*/)
{
   QFile inputFile((filePath == "") ? ":/input.txt" : filePath);
   inputFile.open(QIODevice::ReadOnly);

   QTextStream in(&inputFile);
   QString lines = in.readAll();
   inputFile.close();

   ui->textEdit->setPlainText(lines);
   QTextCursor cursor = ui->textEdit->textCursor();
   cursor.movePosition(QTextCursor::Start, QTextCursor::MoveAnchor, 1);
}

void TextFinder::onFindButtonClicked()
{
   // move cursor to start of file
   QTextCursor cursor = ui->textEdit->textCursor();
   cursor.movePosition(QTextCursor::Start, QTextCursor::MoveAnchor, 1);
   QString searchString = ui->findText->text();
   ui->textEdit->find(searchString, QTextDocument::FindWholeWords);
}

void TextFinder::onOpenButtonClicked()
{
   const QStringList docsLocation = QStandardPaths::standardLocations(
      QStandardPaths::DocumentsLocation);
   QString fileName = QFileDialog::getOpenFileName(
      this, tr("Open File"), docsLocation.last(),
      tr("Text Files (*.txt *.c *.cpp *.h *.hxx *.py *.java *.bat *.sh)"));
   if (fileName != "")
      loadTextFile(fileName);
}

File: bogo2bogo/TextFinder/TextFinder.py
# -*- coding: utf-8 -*-
# textFinder.py - based on Online example from Qt Documentation
import sys
import os
import pathlib
from PyQt5.QtCore import *
from PyQt5.QtGui import *
from PyQt5.QtWidgets import *
from PyQt5 import uic
# to detect dark themes (@see: https://pypi.org/project/darkdetect/)
import darkdetect

sys.path.append(os.path.join(pathlib.Path(__file__).parents[1], 'common_files'))
import mypyqt5_utils as utils

import TextFinderForm


class TextFinder(QWidget):
    def __init__(self):
        super(TextFinder, self).__init__()
        # self.setWindowTitle(f'Text Finder Demo: PySide {PySide2.__version__}')
        self.setWindowTitle(f'Text Finder Demo: PyQt {PYQT_VERSION_STR}')
        p = pathlib.Path(__file__)
        uiFilePath = os.path.join(os.path.split(str(p))[0], "TextFinder.ui")
        # uiFilePath = os.path.join(str(p.parent[1]), "TextFinder.ui")
        self.ui = uic.loadUi(uiFilePath, self)
        # self.ui = uiLoader.load(uiFilePath, self)
        self.loadTextFile()
        self.ui.textEdit.setEnabled(False)
        self.ui.findButton.setMinimumWidth(100)
        self.ui.findButton.clicked.connect(self.findButtonClicked)

    def loadTextFile(self):
        p = pathlib.Path(__file__)
        filePath = os.path.join(os.path.split(str(p))[0], "input.txt")
        with open(filePath, 'r') as f:
            lines = ''.join(f.readlines())  # convert list to str
            # lines = ''.join(lines)
            self.ui.textEdit.setText(lines)

    def findButtonClicked(self):
        # move text cursor to beginning of text file
        cursor = self.ui.textEdit.textCursor()
        cursor.movePosition(QTextCursor.Start, QTextCursor.MoveAnchor, 1)
        searchString = self.ui.findText.text()
        self.ui.textEdit.find(searchString)  # , QTextDocument.FindWholeWords)


def main():
    app = QApplication(sys.argv)
    font = QFont("SF UI Text", QApplication.font("QMenu").pointSize())
    app.setFont(font)  # QApplication.font("QMenu"))
    app.setStyle("Fusion")
    palSwitcher = utils.PaletteSwitcher(app)

    if darkdetect.isDark():
        #utils.setDarkPalette(app)
        palSwitcher.setDarkPalette()

    w = TextFinder()
    w.setFont(font)
    w.show()

    return app.exec_()


if __name__ == "__main__":
    main()
